src/idl/src/tree.c

```diff
@@ -983,6 +984,204 @@ static const char *describe_struct(const void *ptr)
+      if (!mod->autoid.annotation) {
+        mod->autoid.value = toset;
+        ret = idl_propagate_autoid(pstate, mod->definitions, toset);
+      }
+    } else if (idl_is_struct(node)) {
+      idl_struct_t *str = (idl_struct_t*)node;
+      if (!str->autoid.annotation) {
+        str->autoid.value = toset;
+        ret = idl_propagate_autoid_struct(pstate, str);
+      }
+    } else if (idl_is_union(node)) {
+      idl_union_t *u = (idl_union_t*)node;
+      if (!u->autoid.annotation) {
+        u->autoid.value = toset;
+        ret = idl_propagate_autoid_union(pstate, u);
+      }
+    }
+    if (ret)
+      break;
+  }
+
+  return ret;
+}
+
+static void
+process_declarator_id(
+  idl_declarator_t *decl,
+  idl_declarator_t *last,
+  idl_autoid_t autoid)
+{
+  assert(decl);
+
+  if (decl->id.annotation) {
+    /*this declarator already has an ID annotation*/
+  } else if (autoid == IDL_HASH) {
+    assert(decl->name->identifier);
+    decl->id.value = idl_hashid(decl->name->identifier);
+  } else if (last) {
+    decl->id.value = (last->id.value + 1) & 0x0FFFFFFFu;
+  } else {
+    decl->id.value = 0;
+  }
+}
+
+idl_retcode_t
+idl_propagate_autoid_struct(
+  idl_pstate_t *pstate,
+  idl_struct_t *str)
+{
+  assert(str);
+
+  size_t n_ids = 0;
+  idl_member_t *mem = NULL;
+  idl_declarator_t *decl = NULL,
+                   *strlast = NULL,
+                   *last = NULL;
+  idl_retcode_t ret = IDL_RETCODE_OK;
+
+  idl_struct_t *current = str;
+  while (current) {
+    IDL_FOREACH(mem, current->members) {
+      IDL_FOREACH(decl, mem->declarators) {
+        n_ids++;
+        if (current != str && !last)
+          strlast = decl;
+      }
+    }
+    if (!last)
+      last = strlast;
+
+    if (current->inherit_spec)
+      current = current->inherit_spec->base;
+    else
+      current = NULL;
+  }
+
+  if (!n_ids)
+    return ret;
+
+  idl_declarator_t **id_array = malloc(sizeof(idl_declarator_t *)*n_ids);
+  if (!id_array) {
+    ret = IDL_RETCODE_NO_MEMORY;
+    goto finish;
+  }
+
+  n_ids = 0;
+  current = str;
+  while (current) {
+    IDL_FOREACH(mem, current->members) {
+      IDL_FOREACH(decl, mem->declarators) {
+        if (current == str)
+          process_declarator_id(decl, last, str->autoid.value);
+
+        last = decl;
+        id_array[n_ids++] = decl;
+      }
+    }
+
+    if (current->inherit_spec)
+      current = current->inherit_spec->base;
+    else
+      current = NULL;
+  }
+
+  qsort(id_array, n_ids, sizeof(*id_array), &compare_declarator);
+
+  for (size_t i = 1; i < n_ids; i++) {
+    if (id_array[i-1]->id.value == id_array[i]->id.value) {
+      idl_error(pstate, idl_location(id_array[i]),
+        "Struct member id '0x%.07x' is already assigned", id_array[i]->id.value);
+      ret = IDL_RETCODE_SEMANTIC_ERROR;
+    }
+  }
+
+  finish:
+  if (id_array)
+    free((void*)id_array);
+
+  return ret;
+}
+
+idl_retcode_t
+idl_propagate_autoid_union(
+  idl_pstate_t *pstate,
+  idl_union_t *u)
+{
```

Comment:
There's quite a bit of duplication between all these functions. How about just writing some iteration functions that can be passed as a function argument? A bit inline with what `idl_iterate` does.

Reply:
There is some code being duplicated here, true.
But the actions between the two functions (for instance, which items they iterate over, unions directly over cases, structs over members and declarators and everything that is inherited) are different enough IMO that creating a more generic implementation would only add complexity, which is why I chose this more explicit approach.

Reply:
The logic is exactly the same. Determine the number of declarators, allocate, assign then sort. So implementing a static function to return the next declarator to iterate is more elegant IMHO. It'll even save complexity if you'd only use it for `struct` types...